clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionSimulationCoordinator.java

```diff
@@ -0,0 +1,316 @@
+            } else if (request.data.producerId() != currentProducerId) {
+                errors.put(topicPartition, Errors.UNKNOWN_PRODUCER_ID);
+            } else if (request.data.producerEpoch() != currentEpoch) {
+                errors.put(topicPartition, Errors.INVALID_PRODUCER_EPOCH);
+            } else {
+                errors.put(topicPartition, Errors.NONE);
+            }
+        });
+
+        return new AddPartitionsToTxnResponse(
+            throttleTimeMs,
+            errors
+        );
+    }
+
+    private AddOffsetsToTxnResponse handleAddOffsetsToTxn(AddOffsetsToTxnRequest request,
+                                                          final boolean faultInject) {
+        if (faultInject) {
+            return new AddOffsetsToTxnResponse(
+                new AddOffsetsToTxnResponseData()
+                    .setErrorCode(Errors.COORDINATOR_LOAD_IN_PROGRESS.code())
+                    .setThrottleTimeMs(throttleTimeMs)
+            );
+        } else if (request.data.producerId() != currentProducerId) {
+            return new AddOffsetsToTxnResponse(
+                new AddOffsetsToTxnResponseData()
+                    .setErrorCode(Errors.UNKNOWN_PRODUCER_ID.code())
+                    .setThrottleTimeMs(throttleTimeMs)
+            );
+        } else if (request.data.producerEpoch() != currentEpoch) {
+            return new AddOffsetsToTxnResponse(
+                new AddOffsetsToTxnResponseData()
+                    .setErrorCode(Errors.INVALID_PRODUCER_EPOCH.code())
+                    .setThrottleTimeMs(throttleTimeMs)
+            );
+        } else {
+            offsetsAddedToTxn = true;
+            return new AddOffsetsToTxnResponse(
+                new AddOffsetsToTxnResponseData()
+                    .setErrorCode(Errors.NONE.code())
+                    .setThrottleTimeMs(throttleTimeMs)
+            );
+        }
+    }
+
+    private AbstractResponse handleTxnCommit(TxnOffsetCommitRequest request,
+                                             final boolean faultInject) {
+        Map<TopicPartition, Errors> errors = new HashMap<>();
+        request.data.topics().forEach(topic -> topic.partitions().forEach(partition -> {
+            TopicPartition key = new TopicPartition(topic.name(), partition.partitionIndex());
+            if (faultInject) {
+                errors.put(key, Errors.COORDINATOR_LOAD_IN_PROGRESS);
+            } else if (request.data.producerId() != currentProducerId) {
+                errors.put(key, Errors.UNKNOWN_PRODUCER_ID);
+            } else if (request.data.producerEpoch() != currentEpoch) {
+                errors.put(key, Errors.INVALID_PRODUCER_EPOCH);
+            } else if (offsetsAddedToTxn) {
+                pendingOffsets.put(key, partition.committedOffset());
+                errors.put(key, Errors.NONE);
+            } else {
+                errors.put(key, Errors.UNKNOWN_TOPIC_OR_PARTITION);
```

Comment:
In this case, the client has send TxnOffsetCommit before the partition has been added to the transaction, which is an illegal state transition. Could we throw an assertion error or something directly to cause the test to fail?